specs/_features/eip6988/beacon-chain.md

```diff
@@ -0,0 +1,127 @@
+# EIP-6988 -- The Beacon Chain
+
+## Table of contents
+
+<!-- TOC -->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Introduction](#introduction)
+- [Helper functions](#helper-functions)
+  - [Misc](#misc)
+    - [Modified `compute_proposer_index`](#modified-compute_proposer_index)
+    - [Modified `get_beacon_proposer_index`](#modified-get_beacon_proposer_index)
+- [Testing](#testing)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This is the beacon chain specification of [EIP-6988](https://github.com/ethereum/EIPs/pull/6988).
+EIP-6988 modifies the protocol to prevent slashed validator from becoming a block proposer.
+
+*Note:* This specification is built upon [Capella](../../capella/beacon-chain.md) and is under active development.
+
+## Helper functions
+
+### Misc
+
+#### Modified `compute_proposer_index`
+
+*Note:* Disallowing slashed validator to become a proposer is the only modification of this function.
+
+```python
+def compute_proposer_index(state: BeaconState, indices: Sequence[ValidatorIndex], seed: Bytes32) -> ValidatorIndex:
+    """
+    Return from ``indices`` a random index sampled by effective balance.
+    """
+    assert len(indices) > 0
+    MAX_RANDOM_BYTE = 2**8 - 1
+    i = uint64(0)
+    total = uint64(len(indices))
+    while True:
+        candidate_index = indices[compute_shuffled_index(i % total, total, seed)]
+        random_byte = hash(seed + uint_to_bytes(uint64(i // 32)))[i % 32]
+        effective_balance = state.validators[candidate_index].effective_balance
+        # [Modified in EIP6988]
+        slashed = state.validators[candidate_index].slashed
+        if effective_balance * MAX_RANDOM_BYTE >= MAX_EFFECTIVE_BALANCE * random_byte and not slashed:
```

Comment:
This actually has a strange side effect --- the shuffling for an epoch is no longer fixed at the start of the epoch as a slashing could remove a validator from a previously selected slot and *replace* them with someone else.

This isn't a deal breaker but likely breaks engineering invariants. We need to discuss with engineers and make sure to have specific tests for this (i.e. predict the epoch's shuffling, then move a coupel of blocks forward with a slashing for the next slot proposer, then show that the proposer changes)

Reply:
> the shuffling for an epoch is no longer fixed at the start of the epoch as a slashing could remove a validator from a previously selected slot and replace them with someone else

Very good point! 👍 

Strictly, the shuffling for an epoch is not fixed before this change was applied, and in the status quo slashing can become a reason of a proposer switch because of its economics component. This change moves this effect from probabilistic to deterministic area and because of that it should be taken with care.

Reply:
> but I'd want it to be a multi-block test to ensure some epoch proposer shuffling cache doesn't become an issue

I've added [`test_proposer_shuffling_changes_within_epoch`](https://github.com/ethereum/consensus-specs/pull/3371/files#diff-229458074872d12664b8ebec667a5ff1cf22df4986fd25ec21ef6f425e80b6a9R61)

Reply:
> Strictly, the shuffling for an epoch is not fixed before this change was applied

but it is due to `effective_balance` only being updated at epoch boundaries! let's discuss on call today

Reply:
We could compute the proposers in the epoch transition and persist the result in the state. Citing the EIP motivation

> The impact of the proposed fix in the case of a single slashing on Ethereum Mainnet is negligible but it becomes significant in the case of correlated slashings. For instance, a correlated slashing of `1/10th` of a validator set can lead to `1/10th` of missed proposals in a number of epochs after the slashing.

This PR reduces the impact to 0 slots of potential missed proposes. With my suggestion it would be reduced to max of 32 slots.

At least in Lodestar and probably in all clients, breaking dependant_root patterns will have a significant impact on the validator client's design.

Reply:
I've attempted to spec out proposer shuffling in state https://github.com/ethereum/consensus-specs/pull/3405. It doesn't seem like a big change after all.

This modification requires beacon state initialization to be finished with a call to `update_proposer_shuffling` like it is done in [`initialize_beacon_state_from_eth1`](https://github.com/ethereum/consensus-specs/pull/3405/files#diff-d470308b98b2d57ebc7e302ea8f235f8f9255e142588f180c4151144dff68682R209) and [`upgrade_to_6988`](https://github.com/ethereum/consensus-specs/pull/3405/files#diff-55793a04843f449fd6234e86fc0f380b79215c54892c5034a8f3a489a5f73ba5R124). I am wondering about other implication that this requirement can have.